xds/internal/client/client.go

```diff
@@ -0,0 +1,56 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package client contains the implementation of the xds client used by xds.
+package client
+
+import "context"
+
+// ServiceUpdate contains update about the service.
+type ServiceUpdate struct{}
+
+// ClusterUpdate contains update about the cluster.
+type ClusterUpdate struct{}
+
+// EndpointUpdate contains update about the endpoint.
+type EndpointUpdate struct{}
+
+// Client is the xds client resolver and balancer use to watch xds updates.
+//
+// All the watch methods don't block.
+type Client struct {
+}
+
+// WatchService watches LRS/RDS/VHDS.
+func (*Client) WatchService(ctx context.Context, target string, callback func(*ServiceUpdate, error)) {
```

Comment:
Thinking about it now, I am not sure if taking a context in these watch apis is going to be useful.
The way I see it, these functions are going to be non-blocking, so they will just save the callback, push something onto some channel and return immediately.
The client will have some long running goroutine which would listen on this channel and send appropriate requests out, and there will be another goroutine which gets the responses and calls the callbacks. These goroutines will select on the top-level context stored in the `Client` object (so calling `Close` on the `Client` should make them exit). Let me know if you had thought of other reasons for having the `context` here.

Also, the registered callbacks for these watch APIs should be non-blocking.